electronic_mail.py

```diff
@@ -62,22 +99,35 @@ def send_mail(cls, mails):
 
             recipients = recipients_from_fields(mail)
 
-            mail_smtp_server = (mail.template.smtp_server or smtp_server
-                if mail.template else smtp_server)
-            mail_draft_mailbox = (mail.template.draft_mailbox or draft_mailbox
-                if mail.template else draft_mailbox)
+            mail_smtp_server = (mail.template.smtp_server if mail.template
+                else smtp_server)
+            mail_draft_mailbox = (mail.template.draft_mailbox if mail.template
+                else draft_mailbox)
+
+            if not mail_smtp_server or not mail_draft_mailbox:
+                if not mail_smtp_server:
+                    logger.warning(
+                        'Missing default SMTP server mail ID: %s' % mail.id)
+                if not mail_smtp_server:
+                    logger.warning(
+                        'Missing default Mailbox mail ID: %s' % mail.id)
+                continue
 
             # Validate recipients to send or move email to draft mailbox
-            if not ElectronicMail.validate_emails(recipients):
+            if not cls.validate_emails(recipients, raise_exception=False):
                 to_draft.extend(([mail], {'mailbox': mail_draft_mailbox}))
                 continue
 
+            # lock before we try to send it because there may be concurrent
+            # process aiming to send them
+            cls.lock([mail])
```

Comment:
I think cls.lock() should be called just after `for mail in mails`. Not only this but we should also ensure we read() again the mail from the database after the lock and check if mail.flag_send is set. Something like:

```
for mail in mails:
    cls.lock([mail])
    # Given that the lock (which is a SELECT FOR UPDAT NOWAIT) guarantees that no concurrent
    # process is updating it, we must read now the flag_send value, and not rely on the value that
    # may be in the cache of the object before the lock, which could have an older value.
    mail = cls(mail.id)
    if mail.flag_send:
        continue
    if not mail.mail_file:
        continue
    ...
```

Reply:
Done ba38157